include/exadg/postprocessor/time_control_statistics.cpp

```diff
@@ -0,0 +1,83 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2021 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+// ExaDG
+#include <exadg/postprocessor/time_control_statistics.h>
+#include <exadg/utilities/print_functions.h>
+
+namespace ExaDG
+{
+TimeControlDataStatistics::TimeControlDataStatistics()
+  : write_preliminary_results_every_nth_time_step(numbers::invalid_timestep)
+{
+}
+
+void
+TimeControlDataStatistics::print(dealii::ConditionalOStream & pcout, bool const unsteady) const
+{
+  time_control_data.print(pcout, unsteady);
+  if(Utilities::is_valid_timestep(write_preliminary_results_every_nth_time_step))
+    print_parameter(pcout,
+                    "Write preliminary results every nth time step",
+                    write_preliminary_results_every_nth_time_step);
+}
+
+void
+TimeControlStatistics::setup(TimeControlDataStatistics const & time_control_data_statistics_in)
+{
+  time_control_data_statistics = time_control_data_statistics_in;
+  time_control.setup(time_control_data_statistics.time_control_data);
+}
+
+bool
+TimeControlStatistics::needs_evaluation(double const           time,
+                                        types::time_step const time_step_number) const
+{
+  if(time_control.needs_evaluation(time, time_step_number))
+    return true;
+  if(write_preliminary_results(time_step_number))
+    return true;
```

Comment:
I think we should not mix the logic like this. It is unclear in my opinion what this function does. Writing of preliminary results should only be invoked if `needs_evaluation()` returned true, so the logic has to be kind of nested. This should also allow to avoid the "double increment problem" described here https://github.com/exadg/exadg/pull/262/files#r1001517833. If you nest the functionality, there is not need to call `needs_evaluation()` a second time, with the side effect of incrementing the counter again.

Reply:
needs_evaluation() should also return true if no result is written, but an evaluation is needed inside the evaluate() function.  I will rewrite some code, but the interface should stay the same in my opinion.

Reply:
The problem with the current realization is that the evaluation of the postprocessing routine (`do_evaluate()`) is invoked whenever `write_preliminary_results()` returns true. But we do not want to have this, which is clearly a side effect. If you need - in addition to the above function - separate if-statements inside `evaluate()`, like

```c++
evaluate()
{
if(time_control.needs_evaluation())
  do_evaluate();
if(write_preliminary_results())
  do_write_results();
}
```
there is no use in defining `TimeControlStatistics::needs_evaluation()` in my opinion. You can write the above two if-statements in `Postprocessor::do_postprocessing()` directly and the logic is much clearer.